File: app/src/main/java/com/hfad/myferma/WriteOff/WriteOffFragment.java
package com.hfad.myferma.WriteOff;


import androidx.fragment.app.Fragment;

import android.app.FragmentTransaction;
import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.AdapterView;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;
import com.google.android.material.textfield.TextInputLayout;
import com.hfad.myferma.MainActivity;
import com.hfad.myferma.R;
import com.hfad.myferma.SalePackage.SaleChartFragment;
import com.hfad.myferma.db.MydbManager;
import com.hfad.myferma.db.MydbManagerMetod;

import java.text.DecimalFormat;


public class WriteOffFragment extends Fragment implements View.OnClickListener {
    private TextView result_text, error;

    private RadioButton radioButton1, radioButton2;
    private RadioGroup radioGroup;
    private TextInputLayout addWriteOffEdit, menu;
    private AutoCompleteTextView animals_spiner;
    private MydbManagerMetod mydbManager;
    private String unit = null;

    private DecimalFormat f;
    private int status = 0;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        mydbManager = new MydbManagerMetod(inflater.getContext());
        View layout = inflater.inflate(R.layout.fragment_write_off, container, false);

        addWriteOffEdit = layout.findViewById(R.id.add_edit);
        addWriteOffEdit.getEditText().setOnEditorActionListener(editorListenerWriteOff);

        radioGroup = (RadioGroup) layout.findViewById(R.id.radioGroup);
        radioButton1 = (RadioButton) layout.findViewById(R.id.radio_button_1);
        radioButton2 = (RadioButton) layout.findViewById(R.id.radio_button_2);


        menu = (TextInputLayout) layout.findViewById(R.id.menu);
        result_text = (TextView) layout.findViewById(R.id.result_text);

        error = (TextView) layout.findViewById(R.id.errorText);

        animals_spiner = (AutoCompleteTextView) layout.findViewById(R.id.animals_spiner);
        animals_spiner.setText("Яйца", false);

        f = new DecimalFormat("0");
        animals_spiner.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                switch (position) {
                    case 0:
                        addWriteOffEdit.setSuffixText("шт.");
                        f = new DecimalFormat("0");
                        result_text.setText(f.format(mydbManager.sumSaleEgg()) + " шт.");
                        break;
                    case 1:
                        addWriteOffEdit.setSuffixText("л.");
                        f = new DecimalFormat("0.00");
                        result_text.setText(f.format(mydbManager.sumSaleMilk()) + " л.");
                        break;
                    case 2:
                        addWriteOffEdit.setSuffixText("кг.");
                        f = new DecimalFormat("0.00");
                        result_text.setText(f.format(mydbManager.sumSaleMeat()) + " кг.");
                        break;
                }
            }
        });
        addWriteOffEdit.setStartIconDrawable(R.drawable.baseline_shopping_bag_24);

        Button addWriteOff = (Button) layout.findViewById(R.id.addWriteOff_button);
        addWriteOff.setOnClickListener(this);

        Button addWriteOffChar = (Button) layout.findViewById(R.id.writeOffChart_button);
        addWriteOffChar.setOnClickListener(this);

        ExtendedFloatingActionButton fab = (ExtendedFloatingActionButton) getActivity().findViewById(R.id.extended_fab);
        fab.setOnClickListener(this);

        //настройка верхнего меню фаб кнопку
        MaterialToolbar appBar = getActivity().findViewById(R.id.topAppBar);
        appBar.setTitle("Мои Списания");

        fab.show();
        fab.setText("Журнал");
        fab.setIconResource(R.drawable.ic_action_book);
        fab.getIcon();

        //Радио переключатель
        radioGroup.setOnCheckedChangeListener((group, checkedId) -> {
            switch (checkedId) {
                case R.id.radio_button_1:
                    status = 0;
                    break;

                case R.id.radio_button_2:
                    status = 1;
                    break;
            }
        });

    return layout;
    }

    @Override
    public void onStart() {
        super.onStart();
        View view = getView();
        if (view != null) {
            mydbManager.open();
            result_text.setText(f.format(mydbManager.sumSaleEgg()) + " шт.");
        }
    }

    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.add_edit:
                addWriteOffEdit.getEditText().setOnEditorActionListener(editorListenerWriteOff);
                break;

            case R.id.addWriteOff_button:
                onClickAddSale(v);
                break;

            case R.id.extended_fab:
                Intent intent = new Intent(getActivity(), WriteOffActivity.class);
                startActivity(intent);
                break;
            case R.id.writeOffChart_button:
                writeOffChart(v);
                break;
        }
    }

    private TextView.OnEditorActionListener editorListenerWriteOff = new TextView.OnEditorActionListener() {
        @Override
        public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
            if (actionId == EditorInfo.IME_ACTION_GO) {
                saleInTable();
            }
            return false;
        }
    };

    public void onClickAddSale(View view) {
        MainActivity activity = new MainActivity();
        saleInTable();
        activity.closeKeyboard();
    }

    public void saleInTable() {
        if (!addWriteOffEdit.getEditText().getText().toString().equals("")) {
            String animalsType = animals_spiner.getText().toString();
            String inputUnitString = addWriteOffEdit.getEditText().getText().toString().replaceAll(",", ".").replaceAll("[^\\d.]", "");;

            // Для ввода целых чисел или дробных
            if (animalsType.equals("Яйца")) {
                if (inputUnitString.contains(".")){
                    addWriteOffEdit.setError("Яйца не могут быть дробными...");
                    addWriteOffEdit.getError();
                    return;
                }
            }

            double inputUnit = Double.parseDouble(inputUnitString);

            unitString(animalsType);

            //убираем ошибку
            addWriteOffEdit.setErrorEnabled(false);

            if (comparison(animalsType, inputUnit)) {
                //проверка, что введены цены на товар
                if (status == 0) {
                    mydbManager.insertToDbWriteOff(animalsType, inputUnit, R.drawable.baseline_cottage_24);
                } else {
                    mydbManager.insertToDbWriteOff(animalsType, inputUnit, R.drawable.baseline_delete_24);
                }
                result_text.setText(f.format(mydbManager.sumSale(animalsType)) + unit);
                addWriteOffEdit.getEditText().getText().clear();
                error.setText("");
                addWriteOffEdit.setEndIconDrawable(R.drawable.baseline_done_24);
                addWriteOffEdit.getEndIconDrawable();
            }
        } else {
            addWriteOffEdit.setError("Введите кол-во!");
            addWriteOffEdit.getError();
        }
    }

    public String unitString(String animals) {
        if (animals.equals("Яйца")) {
            unit = " шт.";
        }
        if (animals.equals("Молоко")) {
            unit = " л.";
        }
        if (animals.equals("Мясо")) {
            unit = " кг.";
        }
        return unit;
    }

    //проверка что не уйдем в минус
    boolean comparison(String animalsType, double inputUnit) {
        if (mydbManager.sumSale(animalsType) - inputUnit < 0) {
            addWriteOffEdit.setError("Нет столько товара на складе");
            addWriteOffEdit.getError();
            return false;
        }
        return true;
    }

    public void writeOffChart(View view){
        WriteOffChartFragment writeOffChartFragment = new WriteOffChartFragment();
        getActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.conteiner, writeOffChartFragment, "visible_fragment")
                .addToBackStack(null)
                .commit();
    }


}

File: app/src/main/java/com/hfad/myferma/menu/WriteOffFragment.kt
package com.hfad.myferma.menu

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.Button
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.textfield.TextInputLayout
import com.hfad.myferma.ManagerMenuPackage.AddManagerFragment
import com.hfad.myferma.Chart.WriteOffChartFragment
import com.hfad.myferma.Settings.InfoFragment
import com.hfad.myferma.MainActivity
import com.hfad.myferma.R
import com.hfad.myferma.Settings.SettingsFragment
import com.hfad.myferma.db.MyConstanta
import com.hfad.myferma.db.MyFermaDatabaseHelper
import java.text.DecimalFormat

class WriteOffFragment : Fragment(), View.OnClickListener {
    private lateinit var resultText: TextView
    private lateinit var myDB: MyFermaDatabaseHelper
    private lateinit var radioButton1: RadioButton
    private lateinit var radioButton2: RadioButton
    private lateinit var radioGroup: RadioGroup
    private lateinit var addWriteOffEdit: TextInputLayout
    private lateinit var animalsSpiner: AutoCompleteTextView

    private var tempList = mutableMapOf<String, Double>()
    private var productList = mutableListOf<String>()
    private var f = DecimalFormat("0.00")
    private var status: Int = 0
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val layout = inflater.inflate(R.layout.fragment_write_off, container, false)
        myDB = MyFermaDatabaseHelper(requireContext())

        productList.clear()

        addArray()
        add()

        // Установка EditText
        addWriteOffEdit = layout.findViewById(R.id.add_edit)
        addWriteOffEdit.editText!!.setOnEditorActionListener(editorListenerWriteOff)

        //Установка Radio
        radioGroup = layout.findViewById<View>(R.id.radioGroup) as RadioGroup
        radioButton1 = layout.findViewById<View>(R.id.radio_button_1) as RadioButton
        radioButton2 = layout.findViewById<View>(R.id.radio_button_2) as RadioButton

        // Установка Текста Todo Status воообще
        resultText = layout.findViewById<View>(R.id.result_text) as TextView

        //Установка спинера
        animalsSpiner = layout.findViewById<View>(R.id.animals_spiner) as AutoCompleteTextView
        animalsSpiner.setText(productList[0], false)

        val product: String = animalsSpiner.text.toString()

        resultText.text = f.format(tempList[product]) + unitString(product)

        animalsSpiner.onItemClickListener =
            OnItemClickListener { parent, view, position, id ->
                val productClick: String = productList[position]
                resultText.text = f.format(tempList[productClick]) +  unitString(productClick)
                addWriteOffEdit.suffixText = unitString(productClick)
                addWriteOffEdit.error = null
                addWriteOffEdit.endIconDrawable = null
                addWriteOffEdit.endIconDrawable
            }
        addWriteOffEdit.setStartIconDrawable(R.drawable.baseline_shopping_bag_24)

        val addWriteOff: Button = layout.findViewById<View>(R.id.addWriteOff_button) as Button
        addWriteOff.setOnClickListener { onClickAddWriteOff() }

        val addWriteOffChar: Button = layout.findViewById<View>(R.id.writeOffChart_button) as Button
        addWriteOffChar.setOnClickListener { moveToNextFragment(WriteOffChartFragment()) }

        //настройка верхнего меню фаб кнопку
        val appBar: MaterialToolbar = requireActivity().findViewById(R.id.topAppBar)
        appBar.title = "Мои Списания"
        appBar.menu.findItem(R.id.magazine).isVisible = true
        appBar.setOnMenuItemClickListener(Toolbar.OnMenuItemClickListener { item: MenuItem ->
            when (item.itemId) {
                R.id.magazine -> moveToNextFragment(AddManagerFragment())
                R.id.more -> {
                    moveToNextFragment(InfoFragment())
                    appBar.title = "Информация"
                }

                R.id.setting -> {
                    moveToNextFragment(SettingsFragment())
                    appBar.title = "Мои настройки"
                }
            }
            true
        })

        //Радио переключатель
        radioGroup.setOnCheckedChangeListener { group: RadioGroup?, checkedId: Int ->
            when (checkedId) {
                R.id.radio_button_1 -> status = 0
                R.id.radio_button_2 -> status = 1
            }
        }
        return layout
    }

    override fun onStart() {
        super.onStart()
        val view: View? = view
        if (view != null) {
            val arrayAdapterProduct = ArrayAdapter(
                requireContext().applicationContext,
                android.R.layout.simple_spinner_dropdown_item,
                productList
            )
            animalsSpiner.setAdapter(arrayAdapterProduct)
        }
    }

    private fun addArray() {
        val cursor = myDB.readAllDataProduct()
        while (cursor.moveToNext()) {
            val product: String = cursor.getString(1)
            productList.add(product)
        }
        cursor.close()
    }

    //Формируем список из БД
    private fun add() {
//TODO Возможно ли сделать все через SQL?
        for (product in productList) {

            val cursor = myDB.selectTableNameAndSumCount(
                MyConstanta.TABLE_NAME,
                MyConstanta.TITLE,
                product,
                MyConstanta.DISCROTION
            )

            if (cursor.count != 0) {

                cursor.moveToNext()
                tempList[product] = cursor.getDouble(1)
                cursor.close()

                val cursorSale =
                    myDB.selectTableNameAndSumCount(
                        MyConstanta.TABLE_NAMESALE,
                        MyConstanta.TITLESale,
                        product,
                        MyConstanta.DISCROTIONSale
                    )

                if (cursorSale.count != 0) {
                    cursorSale.moveToNext()
                    tempList[product] = tempList[product]!! - cursorSale.getDouble(1)
                }
                cursorSale.close()

                val cursorWriteOff = myDB.selectTableNameAndSumCount(
                    MyConstanta.TABLE_NAMEWRITEOFF,
                    MyConstanta.TITLEWRITEOFF,
                    product, MyConstanta.DISCROTIONSWRITEOFF
                )

                if (cursorWriteOff.count != 0) {
                    cursorWriteOff.moveToNext()
                    tempList[product] = tempList[product]!! - cursorWriteOff.getDouble(1)
                }
                cursorWriteOff.close()


            } else {
                tempList[product] = 0.0
            }
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.add_edit -> addWriteOffEdit.editText!!
                .setOnEditorActionListener(editorListenerWriteOff)
        }
    }

    private val editorListenerWriteOff: OnEditorActionListener =
        OnEditorActionListener { v, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_GO) {
                saveInTable()
            }
            false
        }

    private fun onClickAddWriteOff() {
        val activity: MainActivity = MainActivity()
        saveInTable()
        activity.closeKeyboard()
    }

    private fun saveInTable() {
        if (addWriteOffEdit.editText!!.text.toString() != "") {

            val animalsType: String = animalsSpiner.text.toString()

            val inputUnitString: String =
                addWriteOffEdit.editText!!.text.toString().replace(",".toRegex(), ".")
                    .replace("[^\\d.]".toRegex(), "")

            // Для ввода целых чисел или дробных
            if (animalsType == "Яйца") {
                if (inputUnitString.contains(".")) {
                    addWriteOffEdit.error = "Яйца не могут быть дробными..."
                    addWriteOffEdit.error
                    return
                }
            }
            val inputUnit: Double = inputUnitString.toDouble()

            //убираем ошибку
            addWriteOffEdit.isErrorEnabled = false

            if (comparison(animalsType, inputUnit)) {
                //проверка, что введены цены на товар
                if (status == 0) {
                    myDB.insertToDbWriteOff(animalsType, inputUnit, R.drawable.baseline_cottage_24)
                } else {
                    myDB.insertToDbWriteOff(animalsType, inputUnit, R.drawable.baseline_delete_24)
                }
                tempList[animalsType] = tempList[animalsType]!! - inputUnit

                resultText.text = f.format((tempList[animalsType])) + unitString(animalsType)

                addWriteOffEdit.editText!!.text.clear()
                addWriteOffEdit.setEndIconDrawable(R.drawable.baseline_done_24)
                addWriteOffEdit.endIconDrawable
            }
        } else {
            addWriteOffEdit.error = "Введите кол-во!"
            addWriteOffEdit.error
        }
    }

    //todo Что-то придумать!
    private fun unitString(animals: String): String {
        return when (animals) {

            "Яйца" -> {
                f = DecimalFormat("0")
                " шт."
            }
            "Молоко" -> " л."
            "Мясо" -> " кг."
            else -> " ед."
        }
    }

    //проверка что не уйдем в минус
    private fun comparison(animalsType: String, inputUnit: Double): Boolean {
        if (tempList[animalsType]!! - inputUnit < 0) {
            addWriteOffEdit.error = "Нет столько товара на складе"
            addWriteOffEdit.error
            return false
        }
        return true
    }

    private fun moveToNextFragment(fragment: Fragment) {
        requireActivity().supportFragmentManager.beginTransaction()
            .replace(R.id.conteiner, fragment, "visible_fragment")
            .addToBackStack(null)
            .commit()
    }
}